### PR TITLE
feat!: create new vault with any keyring

### PR DESCRIPTION
**Pull request by gantunesr**
## Description

This change updates the logic inside the `eth-keyring-controller` to allow the creation of a vault with any keyring. This removes the hardcoded HD Keyring from the code and allow the client to pass the parameters to instantiate any keyring that is accepted by the controller.

## Changes

- **BREAKING**: Unify `createNewVaultAndKeychain` and `createNewVaultAndRestore` into new method `createNewVaultWithKeyring`. `createNewVaultWithKeyring` accepts a password and a keyring object provided by the client and returns the `KeyringControllerState`

## References

No references

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation for new or updated code as appropriate (note: this will usually be JSDoc)
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

File: src/KeyringController.ts [hunk 1]
```diff
@@ -115,60 +115,33 @@ class KeyringController extends EventEmitter {
...
-   * @param password - The password to encrypt the vault with.
-   * @param seedPhrase - The BIP39-compliant seed phrase,
-   * either as a string or Uint8Array.
-   * @returns A promise that resolves to the state.
-   */
-  async createNewVaultAndRestore(
-    password: string,
-    seedPhrase: Uint8Array | string | number[],
+    keyring: {
+      type: string;
```
Comment (owencraston) [inline src/KeyringController.ts:134]: is there now better way we can type the options here?

Comment (owencraston) [reply src/KeyringController.ts:134]: should we default to HD if one is not passed in?

Comment (gantunesr) [reply src/KeyringController.ts:134]: > _is there now better way we can type the options here? I see in the tests we are using KeyringType?_


No, some keyrings are passed from the outside (`QRKeyring`, `LedgerLeyring`, `SnapKeyring`, etc) and there isn't a way to know what they would be unless we enforce it, for the moment having them as `unknown` and having the keyring itself handle it is the best choice.

> _should we default to HD if one is not passed in?_

I don't see the benefit of it at least inside this package, the clients should handle it

**Review by mikesposito — APPROVED** [review]
Looks good!

Question: will this be done in `@metamask/keyring-controller` too?

**Comment by gantunesr** [comment]
@mikesposito Yes! I'll have this done in @metamask/keyring-controller too

**Pull request merged**